testenv/gcp/utils.go

```diff
@@ -0,0 +1,70 @@
+// Copyright © 2023 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"cloud.google.com/go/compute/apiv1/computepb"
+	"context"
+	"fmt"
+	"github.com/openclarity/vmclarity/testenv/utils"
+	dockerhelper "github.com/openclarity/vmclarity/testenv/utils/docker"
+)
+
+func (e *GCPEnv) afterSetUp(ctx context.Context) error {
+	req := &computepb.GetInstanceRequest{
+		Project:  ProjectID,
+		Zone:     Zone,
+		Instance: "vmclarity-" + e.envName + "-server",
+	}
+	server, err := e.instancesClient.Get(ctx, req)
+	e.serverIP = server.NetworkInterfaces[0].AccessConfigs[0].NatIP
+
+	e.sshPortForwardInput = &utils.SSHForwardInput{
+		PrivateKey:    e.sshKeyPair.PrivateKey,
+		User:          DefaultRemoteUser,
+		Host:          *e.serverIP,
+		Port:          utils.DefaultSSHPort,
+		LocalPort:     8080, //nolint:gomnd
+		RemoteAddress: "localhost",
+		RemotePort:    80, //nolint:gomnd
+	}
+
+	e.sshPortForward, err = utils.NewSSHPortForward(e.sshPortForwardInput)
+	if err != nil {
+		return fmt.Errorf("failed to setup SSH port forwarding: %w", err)
+	}
+
+	if err = e.sshPortForward.Start(context.Background()); err != nil { //nolint:contextcheck
```

Comment:
Okay so, we have a problem with passing contexts to these functions from the parent.

The issue why we can't use the `ctx` variable here from the `afterSetup` function's parameter is because if you follow up the call chain then that `ctx` is the `setupCtx` variable in the `beforeSuit` function, see here: https://github.com/openclarity/vmclarity/blob/main/e2e/suite_test.go#L68

That context ends when the `beforeSuit` function runs to completion and after that the context is not available, but with the current solution we want to use that context in actual tests (because we want to have an open SSH connection to VMClarity control plane while the tests are running).

This is not going to work, we have to rework how we pass context from ginkgo's test suite creation to the actual test environments, but I think that's out of the scope of this PR.

The situation is the same here in your previous comment - https://github.com/openclarity/vmclarity/pull/1601/files/de050880ac466a03a92bbf25a6497d4bde060beb#diff-7be911ff144b8813d4374c3f85695689dd505e9ace576d90e7f0f2b7a071b421R327 - the context that we pass to the `iamService` is cancelled when the `beforeSuit` ends, but we want to use this context in the `TearDown` where we want to undelete the 2 roles.